### PR TITLE
node: implement tendermint modes

### PR DESCRIPTION
**Pull request by cmwaters**
Reference PR: #6061

### CODE REVIEW

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -51,6 +51,8 @@ Friendly reminder: We have a [bug bounty program](https://hackerone.com/tendermi
 
 ### FEATURES
 
+- [config] Add `--mode` flag and config variable. See [ADR-52](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-052-tendermint-mode.md)
```
Comment (marbar3778) [inline CHANGELOG_PENDING.md:54]:
```suggestion
- [config] Add `--mode` flag and config variable. See [ADR-52](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-052-tendermint-mode.md) @dongsam
```

File: node/node.go [hunk 2]
```diff
@@ -313,13 +325,18 @@ func logNodeStartupInfo(state sm.State, pubKey crypto.PubKey, logger, consensusL
...
-	} else {
-		consensusLogger.Info("This node is not a validator", "addr", addr, "pubKey", pubKey)
+	switch {
+	case mode == cfg.ModeFull:
+		consensusLogger.Info("This node is a fullnode")
+	case mode == cfg.ModeValidator:
+		addr := pubKey.Address()
+		// Log whether this node is a validator or an observer
+		if state.Validators.HasAddress(addr) {
+			consensusLogger.Info("This node is a validator", "addr", addr, "pubKey", pubKey)
```
Comment (marbar3778) [inline node/node.go:335]: Nit:

```suggestion
			consensusLogger.Info("This node is a validator", "addr", addr, "pubKey", pubKey.Bytes())
```
or String().

**Review by marbar3778 — APPROVED** [review]
Thanks for pushing this over the finish line!!

**Comment by cmwaters** [comment]
There are a few points / ideas that came to mind whilst working on this that I would like to document (I may do also in a separate issue). This PR (and the prior work and ADR) ultimately move us in the direction of making clearer separations in the types of nodes and the roles they play in the network. Specifically:
- Validator Node -> proposes and votes on blocks
- Full Node -> gossips txs, participates in consensus, validates and stores blocks, provides endpoints for data availability
- Seed Node -> keeps an address book of a large part / the entire network, providing peers that nodes can connect with.

I find it compelling to add a fourth type of node:
- Light Node -> runs a light client and an RPC client. 

This is something that of course already exists but I believe needs to be explicitly named in a way that groups it with the rest of the nodes and is more intuitive for people familiarising themselves with Tendermint / Cosmos SDK. To make the distinction clear:
- IBC uses a light client
- A wallet runs a light node

The next point I would like to address is the `Config` file. Given this growing separation between the different nodes in a network, I see two paths we can go down: 1) A "one config to rule them all" approach where one large config file can be used for every node. 2) A config file specific to the node. Currently we side with the former. What this means is that an operator could stop a full node and run a seed node from the exact config. It also means that If I'm running a seed node, my `config.toml` file is filled with variables that have no relation to a seed node i.e. `MempoolConfig` or `ConsensusConfig`. I think this is just bad UX and may confuse end users. I would also argue that it would be more beneficial to this separation if an operator couldn't simply switch modes but would have to start a new process.

I would propose that all node types share a `BaseConfig`(this already exists). The `BaseConfig` would indicate the mode and thus

**Comment by marbar3778** [comment]
> I find it compelling to add a fourth type of node:
> 
>     * Light Node -> runs a light client and an RPC client.

I would say this is more up to the application (https://github.com/cosmos/cosmos-sdk/issues/6563). Reason we can not offer this type of node is because we don't know the apps proof design making queries through the light client not usable. IMO we shouldn't have a cmd for it either, it should be app specific. 

I agree with your second point. Having two configs will cause less confusion and better the UX for outr users.

**Comment by cmwaters** [comment]
> I would say this is more up to the application (cosmos/cosmos-sdk#6563). Reason we can not offer this type of node is because we don't know the apps proof design making queries through the light client not usable. IMO we shouldn't have a cmd for it either, it should be app specific.

Right, the verification of ABCI queries is a difficult one although I believe any app through the cosmos sdk has the same proof design which is the default of the TM light node. Also, aren't the other uses of a light node, verification of headers and txs also valuable?

**Comment by marbar3778** [comment]
> Right, the verification of ABCI queries is a difficult one although I believe any app through the cosmos sdk has the same proof design which is the default of the TM light node. Also, aren't the other uses of a light node, verification of headers and txs also valuable?

Does it work? Last I checked it was still broken. What is the use case for verifying headers via cli? Verifying txs is possible, but it's more valuable to have a single light client defined by the app to do both queries and tx verification.  We can add it, but I would guess it would cause confusion with so many options.

**Comment by marbar3778** [comment]
> > Right, the verification of ABCI queries is a difficult one although I believe any app through the cosmos sdk has the same proof design which is the default of the TM light node. Also, aren't the other uses of a light node, verification of headers and txs also valuable?
> 
> Does the light client query system work? Last I checked it was still broken. What is the use case for verifying headers via cli? Verifying txs is possible, but it's more valuable to have a single light client defined by the app to do both queries and tx verification. We can add it, but I would guess it would cause confusion with so many options.

**Comment by cmwaters** [comment]
> Last I checked it was still broken.

Maybe, I thought we fixed it. At least there is a test for it. 😬 

> What is the use case for verifying headers via cli?

Well technically it's done via the RPC 

> more valuable to have a single light client defined by the app to do both queries and tx verification

Totally agree. I think I need to do some more thinking about it because I would like to make light clients/nodes something that provides greater utility to applications. Currently, the model is to have light clients quite separate from application. The light client doesn't store any app state or have any connection with the app. Perhaps we could have an _ABCI Light_ (😅 ) that gives the application the app hash and a proof for verification. 

I think a valuable use case for a light node would be to track partial state. I.e. perhaps I just want to know all the transactions that occur with a specific type or attribute (a logical example would be tracking transactions to and from an account). A light node could listen for events on a full node and when the tx was made, verify it and update that specific part of state. I don't know, I'm just thinking aloud.

**Comment by marbar3778** [comment]
> Totally agree. I think I need to do some more thinking about it because I would like to make light clients/nodes something that provides greater utility to applications. Currently, the model is to have light clients quite separate from application. The light client doesn't store any app state or have any connection with the app. Perhaps we could have an _ABCI Light_ (😅 ) that gives the application the app hash and a proof for verification.
> 
> I think a valuable use case for a light node would be to track partial state. I.e. perhaps I just want to know all the transactions that occur with a specific type or attribute (a logical example would be tracking transactions to and from an account). A light node could listen for events on a full node and when the tx was made, verify it and update that specific part of state. I don't know, I'm just thinking aloud.

I like some of these ideas. We should do a brainstorm session on these and other ideas, the scope can easily get large, but it would be fun to see what we come up with.

File: node/node.go [hunk 3]
```diff
@@ -813,6 +830,101 @@ func startStateSync(ssR *statesync.Reactor, bcR fastSyncReactor, conR *cs.Reacto
...
+	p2pLogger := logger.With("module", "p2p")
+	transport := createTransport(p2pLogger, config)
+	sw := createSwitch(
+		config, transport, p2pMetrics, nil, nil,
+		nil, nil, nil, nil, nodeInfo, nodeKey, p2pLogger,
+	)
+
+	err = sw.AddPersistentPeers(splitAndTrimEmpty(config.P2P.PersistentPeers, ",", " "))
+	if err != nil {
+
```
Comment (tychoish) [inline node/node.go:866]: nit, blank line

File: node/node.go [hunk 4]
```diff
@@ -1164,31 +1284,34 @@ func (n *Node) OnStart() error {
 		return err
 	}
 
-	if n.config.FastSync.Version == "v0" {
-		// Start the real blockchain reactor separately since the switch uses the shim.
-		if err := n.bcReactor.Start(); err != nil {
-			return err
+	if n.config.Mode != cfg.ModeSeed {
+
```
Comment (tychoish) [inline node/node.go:1288]: nit: line

File: node/node.go [hunk 5]
```diff
@@ -1164,31 +1284,34 @@ func (n *Node) OnStart() error {
...
+		if n.config.FastSync.Version == "v0" {
```
Comment (tychoish) [inline node/node.go:1289]: is "v0" a constant somewhere that we could use?

File: node/node.go [hunk 6]
```diff
@@ -1337,7 +1458,15 @@ func (n *Node) ConfigureRPC() error {
 		Logger: n.Logger.With("module", "rpc"),
 
 		Config: *n.config.RPC,
-	})
+	}
+	if n.config.Mode == cfg.ModeValidator {
+		pubKey, err := n.privValidator.GetPubKey(context.TODO())
+		if pubKey == nil || err != nil {
+			return fmt.Errorf("can't get pubkey: %w", err)
```
Comment (tychoish) [inline node/node.go:1465]: is this error salient when pubkey is nil and error is nil?

File: node/node.go [hunk 7]
```diff
@@ -1630,6 +1759,45 @@ func makeNodeInfo(
...
+	lAddr := config.P2P.ExternalAddress
+
+	if lAddr == "" {
+		lAddr = config.P2P.ListenAddress
+	}
+
+	nodeInfo.ListenAddr = lAddr
+
+	err := nodeInfo.Validate()
+	return nodeInfo, err
```
Comment (tychoish) [inline node/node.go:1798]: should we return a non-nil node info when err is non nil?

File: node/node_test.go [hunk 8]
```diff
@@ -522,6 +522,27 @@ func TestNodeNewNodeCustomReactors(t *testing.T) {
 	assert.Equal(t, customBlockchainReactor, n.Switch().Reactor("BLOCKCHAIN"))
 }
 
+func TestNodeNewSeedNode(t *testing.T) {
+	config := cfg.ResetTestRoot("node_new_node_custom_reactors_test")
+	config.Mode = cfg.ModeSeed
+	defer os.RemoveAll(config.RootDir)
```
Comment (tychoish) [inline node/node_test.go:528]: I'm not sure, but it seems like t.Cleanup() might be better at cleanup but I'm not sure.

File: node/node.go (hunk 5 above)
Comment (cmwaters) [reply node/node.go:1289]: Nope, I don't think so. We could add one

File: node/node.go (hunk 6 above)
Comment (cmwaters) [reply node/node.go:1465]: yup. I'm not sure if there is a case where it happens, but if it does we should still error

File: node/node.go (hunk 7 above)
Comment (cmwaters) [reply node/node.go:1798]: The assumption is generally to read the error first, so I think it should be fine

**Pull request merged**